daffodil-runtime1/src/main/scala/org/apache/daffodil/udf/UserDefinedFunctionService.scala

```diff
@@ -0,0 +1,185 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.daffodil.udf
+
+import collection.JavaConverters._
+import collection.mutable._;
+import java.util.ServiceLoader
+import java.util.ServiceConfigurationError
+import org.apache.daffodil.util.Misc
+import java.io.Serializable
+import org.apache.daffodil.util.LogLevel
+import org.apache.daffodil.util.Logging
+import java.lang.reflect.Method
+
+/**
+ * Loads all User Defined Function Providers on the classpath once and caches them in
+ * following look up tables that'll be used to look up the UDFs during Schema compile time.
+ *
+ * "$namespaceURI:$fName" -> classOf[UserDefinedFunction]
+ * "$namespaceURI:$fName" -> UserDefinedFunctionProvider
+ *
+ * It also does validation on the providers and classes and caches any warning and errors,
+ * which are then reported during compile time.
+ *
+ */
+object UserDefinedFunctionService extends Logging {
+  val evaluateMethodName = "evaluate"
+  type EvaluateMethod = Method
+  type ParameterTypes = Array[Class[_]]
+  type ReturnType = Class[_]
+  type UdfCallingInfo = (UserDefinedFunction, EvaluateMethod, ParameterTypes, ReturnType)
+  type NamespaceNameKey = String
+  private val functionClassesLookup: HashMap[NamespaceNameKey, Class[_]] = HashMap()
+  private val functionProviderLookup: HashMap[NamespaceNameKey, UserDefinedFunctionProvider] = HashMap()
+  private val InitializedUserDefinedFunctionsCache: HashMap[NamespaceNameKey, Option[UdfCallingInfo]] = HashMap()
+  lazy val classUserDefinedFunctionInfo = classOf[UserDefinedFunctionInfo]
+  lazy val classUserDefinedFunction = classOf[UserDefinedFunction]
+  lazy val classJSerializable = classOf[java.io.Serializable]
+  lazy val currentClassPath = Misc.classPath.map(_.toString).mkString("\n")
+
+  val loader: ServiceLoader[UserDefinedFunctionProvider] = ServiceLoader.load(classOf[UserDefinedFunctionProvider])
+
+  loader.asScala.map { provider =>
+    val providerfc = provider.getUserDefinedFunctionClasses
+    lazy val providerClassName = provider.getClass.getName
+
+    if (providerfc == null || providerfc.isEmpty) {
+      val wstr = s"User Defined Function Provider ignored: ${providerClassName}. No User Defined Functions found."
+      log(LogLevel.Warning, wstr)
+    } else {
+      val missingAnnotations = providerfc.filterNot(_.isAnnotationPresent(classUserDefinedFunctionInfo))
+      val nonUserDefinedFunction = providerfc.filterNot(classUserDefinedFunction.isAssignableFrom(_))
+
+      if (!missingAnnotations.isEmpty) {
+        val wstr = s"User Defined Function Provider ignored: ${providerClassName}. Below must be annotated with ${classUserDefinedFunctionInfo.getName}" +
```

Comment:
We actually ignore the entire provider if it has a broken UDF. So it ok to log one line for "UDFP ignored" and then a log for each problematic UDF?
i.e 
```scala
        log(..., "User Defined Function Provider ignored: %s", provider)
        missingAnnotations.foreach { log(..., "Missing %s Annotation: %s",  udfinfo, udf) }
```

Reply:
Ah I see https://github.com/apache/incubator-daffodil/pull/273#discussion_r334920717 answers my question, will ignore only broken UDFs rather than the entire UDFP